packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -2641,22 +2642,37 @@ export class ContainerRuntime
 			}
 			const batchStartCsn = processResult.batchStartCsn;
 			const batch = processResult.messages;
+			const sequenceNumber = processResult.sequenceNumber;
 			const messages: {
 				message: InboundSequencedContainerRuntimeMessage;
 				localOpMetadata: unknown;
+				sequenceNumber?: number;
 			}[] = local
-				? this.pendingStateManager.processPendingLocalBatch(batch, batchStartCsn)
+				? this.pendingStateManager.processPendingLocalBatch(
+						batch,
+						batchStartCsn,
+						sequenceNumber,
+					)
 				: batch.map((message) => ({ message, localOpMetadata: undefined }));
-			messages.forEach(({ message, localOpMetadata }) => {
-				const msg: MessageWithContext = {
-					message,
-					local,
-					modernRuntimeMessage,
-					savedOp,
-					localOpMetadata,
-				};
-				this.ensureNoDataModelChanges(() => this.processCore(msg));
-			});
+			if (messages.length === 0) {
+				assert(sequenceNumber !== undefined, "sequenceNumber must be defined");
+				this.emit("batchBegin", { sequenceNumber });
```

Comment:
Looking at the ScheduleManager code (which owns emitting these events), I'm not sure why you need to do this here as well. There must have been some test that failed or something?

Here's why I think it should work - The empty batch is represented by a single message, and that message should hit the ScheduleManager. ScheduleManager handles single-message-batches properly.

Can you explain why that wasn't happening?

Reply:
Maybe we could recreate a single message and bypass `validateAndProcessRuntimeMessage` and emmiting 'op'